sklearn/neighbors/base.py

```diff
@@ -99,6 +98,172 @@ def _get_weights(dist, weights):
                          "'distance', or a callable function")
 
 
+def _is_sorted_by_data(graph):
+    """Returns whether the graph's non-zero entries are sorted by data
+
+    The non-zero entries are stored in graph.data and graph.indices.
+    For each row (or sample), the non-zero entries can be either:
+        - sorted by indices, as after graph.sort_indices()
+        - sorted by data, as after _check_precomputed(graph)
+        - not sorted.
+
+    Parameters
+    ----------
+    graph : CSR sparse matrix, shape (n_samples, n_samples)
+        Neighbors graph as given by kneighbors_graph or radius_neighbors_graph
+
+    Returns
+    -------
+    res : boolean
+        Whether input graph is sorted by data
+    """
+    assert issparse(graph)
+    out_of_order = graph.data[:-1] > graph.data[1:]
+    return (out_of_order.sum() == out_of_order.take(graph.indptr[1:-1] - 1,
```

Comment:
Rather than subtract 1, should we just take from `out_of_order[1:]`?

Reply:
No it's the opposite, `out_of_order[1:]` is equivalent to `+1`:
```py
np.arange(100).take([10], mode='clip')  # 10
np.arange(100).take([10 - 1], mode='clip')  # 9
np.arange(100)[1:].take([10], mode='clip')  # 11
```